File: HW2-PL/program1.ml
(* Define the types *)
datatype Gender = M | F
datatype Relationship = Parent | Siblings | Brother | Sister | Cousin

(* Define the person *)
type person = string * Gender * string * string * string

(* Define the relationships *)
fun parent ((_, _, par1, _, _), (_, _, par2, _, _)) = par1 = par2
fun siblings ((_, _, _, sib1, _), (_, _, _, sib2, _)) = sib1 = sib2
fun brother ((name1, M, _, _, _), (name2, M, _, _, _)) = siblings((name1, M, "", "", ""), (name2, M, "", "", ""))
fun sister ((name1, F, _, _, _), (name2, F, _, _, _)) = siblings((name1, F, "", "", ""), (name2, F, "", "", ""))
fun cousin ((_, _, _, _, cousin1), (_, _, _, _, cousin2)) = cousin1 = cousin2

(* Query the relationship *)
fun query_relationship (Person1, Person2, Relationship) =
        case Relationship of
                Parent => parent(Person1, Person2)
            | Siblings => siblings(Person1, Person2)
            | Brother => brother(Person1, Person2)
            | Sister => sister(Person1, Person2)
            | Cousin => cousin(Person1, Person2)

(* List of persons *)
val persons = ref []

(* Add a person *)
fun add_person (p : person) = persons := p :: !persons

val _ = add_person(("Andy", M, "","Bob",""))
val _ = add_person(("Bob", M, "Andy","Cecil","Helen"))
val _ = add_person(("Cecil", M, "Bob","Dennis",""))
val _ = add_person(("Dennis", M, "Cecil","Edward","Pattie"))
val _ = add_person(("Edward", M, "Dennis","Felix",""))
val _ = add_person(("Felix", M, "Edward","",""))
val _ = add_person(("Martin", M, "","Nancy","Gigi"))
val _ = add_person(("Oscar", M, "Nancy","Pattie",""))
val _ = add_person(("Quinn", M, "Pattie","Rebecca",""))
val _ = add_person(("Gigi", F, "","Helen","Martin"))
val _ = add_person(("Helen", F, "Gigi","Iris","Bob"))
val _ = add_person(("Iris", F, "Helen","Jane",""))
val _ = add_person(("Jane", F, "Iris","Kate",""))
val _ = add_person(("Kate", F, "Jane","Liz",""))
val _ = add_person(("Liz", F, "Kate","",""))
val _ = add_person(("Nancy", F, "Martin","Oscar",""))
val _ = add_person(("Pattie", F, "Oscar","Quinn","Dennis"))
val _ = add_person(("Rebecca", F, "Quinn","",""))

(* Test cases *)
val _ = print (Bool.toString (query_relationship(("Bob", M, "","Cecil","Helen"), ("Cecil", M, "Bob","Dennis",""), Parent)) ^ "\n")
val _ = print (Bool.toString (query_relationship(("Helen", F, "Gigi","Iris","Bob"), ("Cecil", M, "Bob","Dennis",""), Parent)) ^ "\n")
val _ = print (Bool.toString (query_relationship(("Helen", F, "Gigi","Iris","Bob"), ("Nancy", F, "Martin","Oscar",""), Sister)) ^ "\n")
val _ = print (Bool.toString (query_relationship(("Edward", M, "Dennis","Felix",""), ("Quinn", M, "Pattie","Rebecca",""), Brother)) ^ "\n")
val _ = print (Bool.toString (query_relationship(("Felix", M, "Edward","",""), ("Rebecca", F, "Quinn","",""), Cousin)) ^ "\n")
val _ = print (Bool.toString (query_relationship(("Cecil", M, "Bob","Dennis",""), ("Iris", F, "Helen","Jane",""), Siblings)) ^ "\n")

File: HW2-PL/program1.pl
% Define a person
% person(Name, Sex, Parent, Child, Spouse).
person('Andy', 'M', '','Bob','').
person('Bob', 'M', 'Andy','Cecil','Helen').
person('Cecil', 'M', 'Bob','Dennis','').
person('Dennis', 'M', 'Cecil','Edward','Pattie').
person('Edward', 'M', 'Dennis','Felix','').
person('Felix', 'M', 'Edward','','').
person('Martin', 'M', '','Nancy','Gigi').
person('Oscar', 'M', 'Nancy','Pattie','').
person('Quinn', 'M', 'Pattie','Rebecca','').
person('Gigi', 'F', '','Helen','Martin').
person('Helen', 'F', 'Gigi','Iris','Bob').
person('Iris', 'F', 'Helen','Jane','').
person('Jane', 'F', 'Iris','Kate','').
person('Kate', 'F', 'Jane','Liz','').
person('Liz', 'F', 'Kate','','').
person('Nancy', 'F', 'Martin','Oscar','').
person('Pattie', 'F', 'Oscar','Quinn','Dennis').
person('Rebrcca', 'F', 'Quinn','','').


% Relation #1: If X and Y are married, and X is Z’s parent, then Y is also Z’s parent.
parent(Parent, Child) :- person(Parent, _, _, Child, _); person(_, _, _, Child, Parent).

% Relation #2: If X is Y’s parent, and X is Z’s parent, then Y and Z are siblings.
siblings(Person1, Person2) :-
    parent(Parent, Person1),
    parent(Parent, Person2),
    Person1 \= Person2.

% Relation #3: If X and Y are siblings, X is male, and Y is male, then X and Y are brothers.
brother(Person1, Person2) :-
    siblings(Person1, Person2),
    person(Person1, 'M', _, _, _),
    person(Person2, 'M', _, _, _).

% Relation #4: If X and Y are siblings, X is female, and Y is female, then X and Y are sisters.
sister(Person1, Person2) :-
    siblings(Person1, Person2),
    person(Person1, 'F', _, _, _),
    person(Person2, 'F', _, _, _).

% Relation #5: If W and X are siblings, W is Y’s parent, and X is Z’s parent, then Y and Z are cousins.
cousin(Person1, Person2) :-
    parent(Parent1, Person1),
    parent(Parent2, Person2),
    siblings(Parent1, Parent2).

% Query the relationship
query_relationship(Person1, Person2, Relationship) :-
    (parent(Person1, Person2), Relationship = parent -> true;
    siblings(Person1, Person2), Relationship = siblings -> true;
    brother(Person1, Person2), Relationship = brother -> true;
    sister(Person1, Person2), Relationship = sister -> true;
    cousin(Person1, Person2), Relationship = cousin -> true;
    false).

% Test cases
query_relationship('Bob', 'Cecil','parent').
query_relationship('Helen', 'Cecil','parent').
query_relationship('Helen', 'Nancy','sister').
query_relationship('Edward', 'Quinn','brother').
query_relationship('Felix', 'Rebecca','cousin').
query_relationship('Cecil', 'Iris','siblings').